### ISSUE TITLE
Evaluate fully local SemIf judgement with optional local reranker

### ISSUE THREAD

**Issue by srobroek** [srobroek/slopvac#162]
## Summary

Evaluate and prototype a **fully local semantic judgement path** for Slopvac using [SemIf](https://github.com/theoleecj/semif), with an optional lightweight local reranker such as [Qwen3-Reranker-0.6B](https://huggingface.co/Qwen/Qwen3-Reranker-0.6B).

The goal is not to replace the current judgement layer by assumption. The goal is to determine whether a local, non-generative decision layer can handle candidate categorisation and a large share of Slopvac's semantic judgement workload while preserving recall, auditability, and the current deterministic host guarantees.

This should be designed so that the same Slopvac integration works:

- entirely locally on a developer workstation;
- in CI against a persistent GPU endpoint;
- against a self-hosted remote SemIf service;
- alongside the current generative judge as a baseline and escalation path.

## Why SemIf is a good fit

SemIf is attractive because its decision mechanism is based on scoring explicit answer alternatives from model logits rather than asking a generative model to emit free-form JSON.

For Slopvac, many existing judgement dimensions are already bounded decisions:

- rule applicability;
- fit: absent / partial / ambiguous / unambiguous;
- harm category;
- repair category;
- warrant level;
- preservation checks;
- confirm / reject routing when derived deterministically from the above.

This maps well to a typed decision interface.

The benchmarked SemIf setup uses stock Qwen weights rather than SemIf-specific fine-tuned weights. The important dependency is therefore the **scoring primitive** — access to the relevant model logits — rather than a proprietary model service.

A local implementation can keep the model resident and expose a very small RPC surface to Slopvac.

## Two semantic roles: validate deterministic lint and execute non-deterministic rules

The same decision-provider path can serve **two distinct roles** in Slopvac.

### 1. Semantic validation of deterministic lint outcomes

Deterministic rules should remain the primary reproducible detector where they are appropriate, but their findings can still be semantically validated.

Example:

```text
deterministic lint rule
        |
        v
candidate finding + exact source span
        |
        v
SemIf
        |
        +-- finding is semantically supported
        +-- finding appears to be a false positive
        +-- ambiguous / escalate
```

This is useful for deterministic checks that deliberately favor recall and therefore produce avoidable false positives. SemIf would not need to rediscover the issue: the deterministic rule already supplies the rule id, candidate span, source offsets, and relevant context. The decision provider only evaluates whether the deterministic trigger corresponds to the semantic defect the rule is intended to detect.

Initially, this validation should be **advisory/auditable** and must not silently alter deterministic exit status or deterministic error/warning counts. Once measured against the gold set and adjudicated corpus, a separate policy decision can determine whether validated semantic rejection should ever suppress or downgrade a deterministic finding.

The report should therefore be able to distinguish at least:

- deterministic finding, semantically confirmed;
- deterministic finding, semantically rejected;
- deterministic finding, semantic validation uncertain/escalated;
- deterministic finding, semantic validation not run.

This gives us a way to measure the real precision of deterministic rules and identify rules whose lexical/structural trigger should be improved rather than adding ad-hoc suppressions.

### 2. Primary execution path for non-deterministic rules

For rules that cannot be expressed reliably as deterministic syntax/structure checks, SemIf can be the primary decision engine.

Example:

```text
candidate passage / relationship
        |
        v
SemIf typed decisions
        |
        +-- applicability
        +-- fit
        +-- harm
        +-- repair
        +-- warrant
        |
        v
deterministic host policy
        |
        +-- CONFIRM
        +-- REJECT
        +-- ABSTAIN / ESCALATE
```

This means the provider abstraction should not be designed merely as a prefilter for today's judgement layer. It should support both:

- **verification mode**: validate an already-fired deterministic rule;
- **judgement mode**: evaluate a genuinely semantic/non-deterministic rule.

Both modes should use the same typed decision machinery, model/revision tracking, cache identity, threshold calibration, and escalation mechanism.

A useful long-term architecture is therefore:

```text
                         parsed document
                               |
                 +-------------+-------------+
                 |                           |
                 v                           v
        deterministic rules          semantic rules
                 |                           |
          candidate findings                 |
                 |                           |
                 v                           |
         SemIf validation <------------------+
                 |
          typed decisions
                 |
                 v
       deterministic host policy
                 |
        reporting / escalation
```

This also gives the evaluation work a second purpose: beyond measuring SemIf as a replacement for generative judgement, it can quantify which existing deterministic rules are genuinely precise, which are high-recall-but-noisy, and where semantic validation adds enough value to justify the extra inference step.

## Proposed architecture

```text
                         source document
                               |
                    deterministic parsing
                               |
                      candidate generation
                               |
                 deterministic admission gates
                               |
                 +---------------------------+
                 |                           |
                 | optional                  |
                 v                           |
        Qwen3-Reranker-0.6B                  |
          high-recall prefilter              |
                 |                           |
                 +-------------+-------------+
                               |
                               v
                         SemIf scorer
                    Qwen3.5-4B initially
                               |
              +----------------+----------------+
              |                |                |
         applicability       fit/harm        preservation
                          repair/warrant        checks
              |                |                |
              +----------------+----------------+
                               |
                               v
                    deterministic host policy
                               |
              +----------------+----------------+
              |                |                |
            REJECT           CONFIRM         UNCERTAIN
                                                 |
                                                 v
                                      existing generative judge
                                      / adjudication path
```

The generative model should remain available initially for:

- uncertain/low-margin cases;
- document-wide or relationship-heavy probe rules;
- evidence discovery where the candidate span is not already known;
- rewrite generation;
- evaluation/adjudication.

## Local deployment

A fully local path should be a first-class target.

Example:

```text
slopvac
   |
   | unix socket / localhost HTTP
   v
semifd
   |
   v
Qwen3.5-4B
   |
   +-- CUDA
   +-- MPS/MLX where supported
   +-- llama.cpp/GGUF where supported
```

The service should keep the model resident rather than reload it for each lint run.

The initial implementation should not hard-code Slopvac to a local process. Slopvac should depend on a small provider contract so the scorer can run locally or remotely without changing judgement logic.

## CI deployment

CI should be able to use the same provider contract against a persistent GPU endpoint:

```text
GitHub Actions runner
        |
        | authenticated HTTPS
        v
    semifd endpoint
        |
        v
  persistent GPU worker
        |
        v
   Qwen3.5-4B
```

This avoids requiring a GPU GitHub runner for every job while retaining self-hosted/local inference.

Desired properties:

- model remains hot between CI jobs;
- endpoint is stateless from Slopvac's perspective;
- requests contain only the text required for the judgement;
- authentication is token/IAM/OIDC-friendly;
- bounded request size and timeout;
- deterministic model/revision recorded in the judgement cache key;
- CI can fail closed or fall back according to explicit configuration.

A later implementation could use a GPU-backed EC2/SageMaker/ECS/EKS/RunPod/etc. service. Slopvac should not depend on the hosting platform.

## Backend contract

Introduce a provider abstraction rather than integrating SemIf directly into judgement policy.

Conceptually:

```python
class DecisionProvider(Protocol):
    def decide(
        self,
        *,
        state: str,
        questions: Mapping[str, DecisionQuestion],
    ) -> DecisionResult:
        ...
```

Question types should cover at least:

- boolean / Noul-style decision;
- finite choice;
- ordered score.

Result data should include:

- selected answer;
- probability distribution or raw normalized scores;
- model id;
- model revision;
- backend/runtime revision;
- inference configuration;
- request digest;
- latency;
- provider-specific metadata needed for reproducibility.

Candidate providers could eventually include:

- `semif-local`;
- `semif-http`;
- TypeSafe Jev;
- djev;
- generative adapter/baseline.

The host must not treat provider probability as Slopvac's existing semantic score or model-authored confidence. It is routing/decision evidence and should remain separate from the Fit/Harm/Repair/Warrant contract.

## Optional local reranker

Do **not** assume a separate reranker is required.

First benchmark SemIf on all admitted candidates.

If candidate volume makes a prefilter worthwhile, test a small local reranker before SemIf. The proposed baseline is:

- `Qwen/Qwen3-Reranker-0.6B`
- permissive Apache-2.0 license;
- small enough to run locally alongside the SemIf model;
- usable as a high-recall rule/candidate relevance filter.

Architecture:

```text
(rule description, candidate passage)
             |
             v
 Qwen3-Reranker-0.6B
             |
      permissive cutoff
             |
       +-----+-----+
       |           |
   prefiltered   SemIf
```

The reranker must optimize for **defect recall**, not precision.

It must not silently remove candidates from coverage. Record a distinct prefilter state and the score/threshold that caused it.

A threshold is acceptable only if evaluation shows essentially no material recall loss.

## Evidence handling

SemIf should not be asked to invent evidence.

For local `SPAN_CANDIDATE` rules, the host already knows the candidate text and source offsets. Prefer:

```text
host:
    proposes exact candidate span

SemIf:
    judges whether the proposed span satisfies rule R
    and returns bounded decisions

host:
    retains source/offset authority
```

This is preferable to asking a generative model to rediscover and requote the same span.

For `PASSAGE_PROBE` rules, evidence discovery is harder. Keep the current generative path initially unless a separate retrieval/span proposal mechanism demonstrates equivalent recall.

## Evaluation

This should ship only after a controlled comparison against the existing judgement infrastructure.

Use:

- existing seeded-defect gold set;
- matched controls;
- human-prose corpus;
- existing majority-Sol adjudication data where applicable.

Compare at minimum:

1. seeded-defect recall;
2. control behavior;
3. false-positive incidence on adjudicated human prose;
4. agreement with majority Sol adjudication;
5. calibration/Brier score where probabilities are available;
6. abstention/escalation rate;
7. candidate reduction if reranking is enabled;
8. total judge calls avoided;
9. latency;
10. local/CI cost.

Required comparisons:

- current generative judge;
- SemIf direct on admitted candidates;
- SemIf with generative escalation for uncertain cases;
- optional reranker -> SemIf;
- no-reranker baseline.

Do not use generic JevBench numbers as acceptance criteria; Slopvac's own corpus is the relevant distribution.

## Thresholding / escalation

Avoid a single global confidence threshold initially.

Calibrate per rule or rule family if the data supports it.

Example policy:

```text
high-margin decision
    -> accept SemIf decision

low-margin / unstable decision
    -> escalate to existing generative judge

probe/evidence-discovery rule
    -> existing judge until independently validated
```

If a reranker is introduced, target approximately >=99% seeded-defect recall before considering candidate elimination useful. The exact threshold must come from measured Slopvac data, not be hard-coded from this issue.

## Security and privacy

A local SemIf deployment has a useful property: source text does not need to leave the developer machine.

For CI with a remote GPU endpoint:

- TLS required;
- authenticate every request;
- no request/response persistence by default;
- redact secrets before semantic judgement using the same source-safety rules as the current judgement path;
- define maximum request size;
- rate limit;
- record model/revision but not sensitive payloads in telemetry;
- allow projects to disable remote semantic judgement entirely.

## Configuration sketch

Something along these lines:

```toml
[judgement.decision_provider]
provider = "semif"
endpoint = "http://127.0.0.1:8941"
model = "Qwen/Qwen3.5-4B"
timeout_seconds = 30

[judgement.reranker]
enabled = false
provider = "local"
model = "Qwen/Qwen3-Reranker-0.6B"

[judgement.escalation]
provider = "existing"
mode = "uncertain"
```

Exact config naming is open.

### 1. Provider abstraction

- [ ] Add a provider-neutral typed decision interface.
- [ ] Add result metadata sufficient for cache/reproducibility.
- [ ] Keep existing judgement output and deterministic gates unchanged.

### 2. SemIf local provider

- [ ] Implement local scorer integration.
- [ ] Support persistent model process/service.
- [ ] Pin model + revision.
- [ ] Include scorer/runtime revision in cache keys.
- [ ] Add deterministic request serialization.

### 3. Evaluation-only mapping

- [ ] Map a subset of local judgement dimensions to typed SemIf questions.
- [ ] Start with local `SPAN_CANDIDATE` rules.
- [ ] Add **verification mode** for already-fired deterministic findings.
- [ ] Add **judgement mode** for non-deterministic/semantic rules.
- [ ] Record confirmed/rejected/uncertain/not-run semantic validation states separately from deterministic outcomes.
- [ ] Do not change production verdict behavior yet.
- [ ] Produce side-by-side evaluation output against the current judge and against deterministic findings.

### 4. Escalation experiment

- [ ] Define margin/uncertainty signal.
- [ ] Route low-margin cases to existing generative judge.
- [ ] Measure recall, precision, escalation rate, cost, and latency.

### 5. Optional reranker experiment

- [ ] Add Qwen3-Reranker-0.6B behind a feature flag.
- [ ] Record every prefiltered candidate.
- [ ] Calibrate thresholds on gold data.
- [ ] Compare SemIf-only vs reranker+SemIf.

### 6. CI remote endpoint

- [ ] Add authenticated HTTP provider.
- [ ] Document persistent GPU deployment.
- [ ] Add CI example/workflow.
- [ ] Verify reproducibility/cache identity across local and remote inference.

## Acceptance criteria

This experiment is successful only if all of the following are demonstrated on Slopvac's own evaluation corpus:

- local judgement runs without an external proprietary inference dependency;
- CI can use the same interface against a remote GPU scorer;
- source/evidence offsets remain host-authoritative;
- the same provider path can validate deterministic findings and execute non-deterministic rules;
- deterministic semantic-validation results are separately reported and auditable without silently changing deterministic pass/fail;
- seeded-defect recall is not materially worse than the existing judge;
- false-positive incidence does not materially regress;
- uncertain cases can be escalated without changing deterministic lint pass/fail semantics;
- model/runtime/revision are fully recorded for reproducibility;
- optional reranking, if retained, demonstrates meaningful work reduction at essentially unchanged recall.

## Non-goals

- Replacing the current judge before evaluation.
- Treating SemIf probabilities as Slopvac confidence scores.
- Allowing semantic judgement to modify deterministic lint findings or exit status.
- Using a reranker as a final defect judge.
- Requiring a cloud service for local Slopvac operation.

## References

- SemIf: https://github.com/theoleecj/semif
- Qwen3-Reranker-0.6B: https://huggingface.co/Qwen/Qwen3-Reranker-0.6B
- Jev model comparison discussed during design: https://benchmarkheaven.com/jev-models
